sentry_streams/src/watermark.rs

```diff
@@ -0,0 +1,144 @@
+use crate::messages::{RoutedValuePayload, WatermarkMessage};
+use crate::routes::{Route, RoutedValue};
+use sentry_arroyo::processing::strategies::{
+    CommitRequest, ProcessingStrategy, StrategyError, SubmitError,
+};
+use sentry_arroyo::types::Message;
+use std::collections::BTreeMap;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+/// Returns the current Unix epoch
+fn current_epoch() -> u64 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_secs()
+}
+
+pub struct Watermark {
+    pub next_step: Box<dyn ProcessingStrategy<RoutedValue>>,
+    pub route: Route,
+    pub period: u64,
+    last_timestamp: u64,
+}
+
+impl Watermark {
+    /// A strategy that periodically sends watermark messages.
+    /// This strategy is added as the first step in a consumer.
+    /// The Arroyo adapter commit step only commits once it recives a watermark
+    /// message with a specific committable from all branches in a consumer.
+    pub fn new(
+        next_step: Box<dyn ProcessingStrategy<RoutedValue>>,
+        route: Route,
+        period: Option<u64>,
+    ) -> Self {
+        let period = match period {
+            Some(p) => p,
+            None => 10,
+        };
+        let last_timestamp = current_epoch();
+        Self {
+            next_step,
+            route,
+            period,
+            last_timestamp,
+        }
+    }
+
+    fn should_send_watermark_msg(&self) -> bool {
+        (self.last_timestamp + self.period) < current_epoch()
+    }
+
+    fn send_watermark_msg(&mut self) -> Result<(), SubmitError<RoutedValue>> {
+        let timestamp = current_epoch();
+        let watermark_msg = RoutedValue {
+            route: self.route.clone(),
+            payload: RoutedValuePayload::WatermarkMessage(WatermarkMessage::new(timestamp as f64)),
+        };
+        self.last_timestamp = timestamp;
+        self.next_step
+            .submit(Message::new_any_message(watermark_msg, BTreeMap::new()))
+    }
+}
+
+impl ProcessingStrategy<RoutedValue> for Watermark {
+    fn poll(&mut self) -> Result<Option<CommitRequest>, StrategyError> {
+        self.next_step.poll()
+    }
+
+    fn submit(&mut self, message: Message<RoutedValue>) -> Result<(), SubmitError<RoutedValue>> {
+        if self.should_send_watermark_msg() {
+            self.send_watermark_msg()?;
+        }
+        self.next_step.submit(message)
```

Comment:
This algorithm will not work. 
If there is no more message passed through we do not send the next watermark. 
The watermark should be sent on poll rather than submit as poll is called periodically no matter if there are new messages from kafka.